CHANGELOG.md

```diff
@@ -1,3 +1,284 @@
+}
+```
+
+**Pass Props to Action**:
+
+```js
+const { withProps } = require('bottender');
+
+async function SayHi(context, { name }) {
+  await context.sendText(`hi! ${name}.`);
+}
+
+async function App() {
+  return withProps(SayHi, { name: 'John' });
+}
+```
+
+**Router**:
+
+```js
+const { router, text } = require('bottender/router');
+
+async function SayHi(context) {
+  await context.sendText('Hi!');
+}
+
+async function SayHello(context) {
+  await context.sendText('Hello!');
+}
+
+async function App() {
+  return router([
+    text('hi', SayHi), // return SayHi when receiving hi text message
+    text('hello', SayHello), // return SayHello when receiving hello text message
+  ]);
+}
+```
+
+**Chain**:
+
+```js
+const { chain } = require('bottender');
+
+function RuleBased(context, props) {
+  if (context.event.text === 'hi') {
+    // discontinue and return SayHi
+    return SayHi;
+  }
+  // continue to next
+  return props.next;
+}
+
+function MachineLearningBased(context, props) {
+  /* ...skip */
+}
+
+function HumanAgent(context, props) {
+  /* ...skip */
+}
+
+function App() {
+  return chain([
+    // will execute in following order
+    RuleBased,
+    MachineLearningBased,
+    HumanAgent,
+  ]);
+}
+```
+
+- [new] Add `_error.js` entry support for error handling:
+
+```js
+// _error.js
+module.exports = async function HandleError(context, props) {
+  await context.sendText(
+    'There are some unexpected errors happened. Please try again later, sorry for the inconvenience.'
+  );
+  console.error(props.error);
+  if (process.env.NODE_ENV === 'production') {
+    // send your error to the error tracker, for example: Sentry
+  }
+  if (process.env.NODE_ENV === 'development') {
+    await context.sendText(props.error.stack);
+  }
+};
+```
+
+- [breaking] `middleware` and Handlers has been moved to `@bottender/handlers` package. You can install it from registry:
+
+```sh
+npm install @bottender/handlers
+
+// or using yarn:
+yarn add @bottender/handlers
+```
+
+And import them like this:
+
+```js
+const {
+  middleware,
+  Handler,
+  MessengerHandler,
+  LineHandler,
+  SlackHandler,
+  TelegramHandler,
+  ViberHandler,
+} = require('@bottender/handlers');
+```
+
+- [breaking] transform all context method parameters to camelcase:
+
+Messenger -
+
+```js
+context.sendGenericTemplate([
+  {
+    title: "Welcome to Peter's Hats",
+    imageUrl: 'https://petersfancybrownhats.com/company_image.png',
+    subtitle: "We've got the right hat for everyone.",
+    defaultAction: {
+      type: 'web_url',
+      url: 'https://peterssendreceiveapp.ngrok.io/view?item=103',
+      messengerExtensions: true,
+      webviewHeightRatio: 'tall',
+      fallbackUrl: 'https://peterssendreceiveapp.ngrok.io/',
+    },
+    buttons: [
+      {
+        type: 'postback',
+        title: 'Start Chatting',
+        payload: 'DEVELOPER_DEFINED_PAYLOAD',
+      },
+    ],
+  },
+]);
+```
+
+Slack -
+
+```js
+await context.chat.scheduleMessage({
```

Comment:
only this one has `await` in example, maybe we can remove it to unify with others